File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    void dfs(vector<vector<int>>& answer, vector<int>& nums, int index, vector<int>& container){
        answer.push_back(container);  
        for(int i = index; i < nums.size(); i++){
            container.push_back(nums[i]); 
            dfs(answer,nums,i+1,container);
            container.pop_back(); 
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> answer; 
        vector<int> container; 
        dfs(answer,nums,0,container);
        return answer; 
    }
};

File: 0078-subsets/0078-subsets.java
class Solution {
    public List<List<Integer>> subsets(int[] nums) {
        List<List<Integer>> answer = new ArrayList<>(); 
        List<Integer> container = new ArrayList<>(); 

        dfs(answer,container,nums,0); 
        return answer; 
    }
    public void dfs(List<List<Integer>> answer, List<Integer> container, int[] nums, int index){
        answer.add(new ArrayList<>(container)); 
        for(int i = index; i < nums.length; i++){
            container.add(nums[i]); 
            dfs(answer,container,nums,i+1); 
            container.removeLast(); 
        }
    }
}